qcodes/instrument/base.py

```diff
@@ -401,6 +410,94 @@ def validate_status(self, verbose: bool = False) -> None:
                     print(f'validate_status: param {k}: {value}')
                 p.validate(value)
 
+    def _add_params_from_decorated_methods(self):
+
+        def special_kwargs_to_meth(self, kwarg_value):
+            """
+            Returns the attribute that has the name `kwarg_value.attr_name`
+            of this instance if the kwarg_value is of type `InstanceAttr`.
+            """
+            if isinstance(kwarg_value, InstanceAttr):
+                return getattr(self, kwarg_value.attr_name)
+            return kwarg_value
+
+        # decorated methods are required to have a specific prefix
+        has_parameter_prefix = lambda s: s.startswith(DECORATED_METHOD_PREFIX)
+
+        for obj_name in filter(has_parameter_prefix, dir(self)):
+            meth = getattr(self, obj_name)
+            # the `@add_parameter` decorator adds an attribute we check for here
+            if hasattr(meth, ADD_PARAMETER_ATTR_NAME):
+                kwargs = {
+                    key: special_kwargs_to_meth(self, val.default)
+                    for key, val in sorted(inspect.signature(meth).parameters.items())
+                }
+                self.add_parameter(
+                    name=meth.__name__[len(DECORATED_METHOD_PREFIX):],
+                    docstring=meth.__doc__,
+                    **kwargs,
+                )
+
+
+class InstanceAttr:
+    """
+    An auxiliary class to be used together with the
+    :obj:`qcodes.instrument.base.add_parameter` decorator to allow adding parameters
+    that require information available only during the `__init__()` of an Instrument
+    subclass.
+    """
+    def __init__(self, attr_name: str):
+        """Instantiates the class and saves the passed in attr_name."""
+        self.attr_name: str = attr_name
+
+    def __repr__(self):
+        """Return the class name and `self.attr_name`."""
+        return f"{self.__class__.__name__}('{self.attr_name}')"
+
+
+DECORATED_METHOD_PREFIX = "_parameter_"
+"""
+A constant defining the prefix of the methods on which
+:obj:`qcodes.instrument.base.add_parameter` decorator can be used. The intention is to
+keep the name of these methods fairly unique and private to avoid any foreseeable clash.
+"""
+
+ADD_PARAMETER_ATTR_NAME = "_add_parameter"
+"""
+A constant defining the name of the attribute set by
+:obj:`qcodes.instrument.base.add_parameter` decorator to flag a method that will be
+converted to parameter.
+"""
+
+def add_parameter(method):
```

Comment:
we use typing a lot in qcodes, so the new methods/classes/functions need to be typed. i think that using a Callable with a type variable (``_T = TypeVar("_T")``) should do the trick
```suggestion
def add_parameter(Callable[_T]) -> Callable[_T]:
```